### DISCUSSION TITLE
How to autocomplete for wrapper multi-languages

### DISCUSSION THREAD

**Discussion by Yokozuna59** [eclipse-langium/langium#1228] [Q&A] [answered]
Hi there,

In our project, we have different languages/diagrams with the same document, so we overridden `ServiceRegistry` and registered each language, and it used a detector to check what language it was.
But in the default case, if no diagram has been detected yet, we want it to autocomplete for possible diagram keywords and snippets, and if any diagram has been detected, it should just switch to it.

![image](https://github.com/eclipse-langium/langium/assets/87622592/39aabd75-0372-4185-9170-095152d03479)

![image](https://github.com/eclipse-langium/langium/assets/87622592/e6f9d652-bdef-4424-a0ad-932cec739ee2)

What is the best way to do so? The shared services don't have `CompletionProvider` service. Do we need to create a separate language that contains all possible diagrams keywords and make it as default value if no diagram detected?

```langium
grammar Mermaid

interface Mermaid {}

entry Mermaid returns Mermaid:
  {Mermaid}
  (
    NEWLINE*
    "info" | "journey" | "mindmap" | "pie" | "quadrant" | "sankey-beta" | "timeline"
    NEWLINE*
  )
;

terminal NEWLINE: /\r?\n/;
// other hidden rules...
```

If that's the case, then how to handle it correctly in `getServices`? It gets glitchy [when trying to match which `uri` match current one](https://github.com/Yokozuna59/multi-lang-example/blob/6ce434356663b5ae981f2416fd9fb68aec1f21de/packages/lsp/src/mermaid/serviceRegistry.ts#L27).

**Comment by msujew** [comment, marked as answer]
> What is the best way to do so? The shared services don't have CompletionProvider service. Do we need to create a separate language that contains all possible diagrams keywords and make it as default value if no diagram detected?

That could work. Note that I would recommend to wrap the keywords in an optional group so they don't automatically generate diagnostics about missing keywords (unless you want that).

> If that's the case, then how to handle it correctly in getServices? It gets glitchy [when trying to match which uri match current one](https://github.com/Yokozuna59/multi-lang-example/blob/6ce434356663b5ae981f2416fd9fb68aec1f21de/packages/lsp/src/mermaid/serviceRegistry.ts#L27).

Can you describe the issues you are experiencing in more detail?

**Reply by Yokozuna59** [reply]
> That could work. Note that I would recommend to wrap the keywords in an optional group so they don't automatically generate diagnostics about missing keywords (unless you want that).

Thanks.

> Can you describe the issues you are experiencing in more detail?

No worries; it was a silly mistake from me. Since I copied the code from `info` diagram, I used `InfoGeneratedModule` in other diagrams services, that's why it keeps using `info` services instead of diagrams ones:

<https://github.com/Yokozuna59/multi-lang-example/blob/741c08f97054f02664e53a5e88a5c4d6a0c43d74/packages/lsp/src/language/mindmap/module.ts#L36>
